### PR TITLE
Convert metabase-lib to TypeScript

### PR DESCRIPTION
**Pull request by ranquild**
Epic: https://github.com/metabase/metabase/issues/19207

Converted metabase-lib with this script https://github.com/metabase/metabase/pull/19205/files#diff-71b379bb2cb007084f9350c35e62ce697ff591b32355e27d16ad07a286b7be44

How to test:
- `yarn lint` should pass
- `yarn test` should pass

### CODE REVIEW

**Comment by alxnddr** [comment]
CI is red 🤔

**Comment by ranquild** [comment]
> CI is red 🤔

@alxnddr yes, we need to merge flow changes first and merge the master afterwards.

**Comment by ranquild** [comment]
> CI is red 🤔

Can be checked now

File: .eslintrc [hunk 1]
```diff
@@ -82,10 +82,6 @@
         "@typescript-eslint/no-unused-vars": [
           "warn",
           { "varsIgnorePattern": "^_" }
-        ],
```
Comment (ranquild) [inline .eslintrc:85]: metabase-lib heavily depends on this

Comment (alxnddr) [reply .eslintrc:85]: Would be nice to enforce `@typescript-eslint/no-use-before-define` it again at some point. Let's maybe have an issue for this?

File: frontend/src/metabase/hoc/AutoExpanding.jsx [hunk 2]
```diff
@@ -9,21 +9,22 @@ import ExplicitSize from "metabase/components/ExplicitSize";
...
-      expand: false,
-    };
-    UNSAFE_componentWillReceiveProps(nextProps) {
-      if (
-        nextProps.width != null &&
-        this.props.width != null &&
-        nextProps.width > this.props.width
-      ) {
-        this.setState({ expand: true });
+  ExplicitSize()(
```
Comment (ranquild) [inline frontend/src/metabase/hoc/AutoExpanding.jsx:12]: TypeScript didn't like this, had to change

File: frontend/src/metabase-lib/lib/metadata/metadata.d.ts [hunk 3]
```diff
@@ -1,36 +0,0 @@
-// to help declaring nominal types
-interface Flavoring<FlavorT> {
```
Comment (alxnddr) [inline frontend/src/metabase-lib/lib/metadata/metadata.d.ts:2]: Why is this one deleted?

Comment (ranquild) [reply frontend/src/metabase-lib/lib/metadata/metadata.d.ts:2]: As Metadata.js is now Metadata.ts, you cannot actually have a .d.ts file for a ts file

File: .eslintrc (hunk 1 above)
Comment (ranquild) [reply .eslintrc:85]: I think the main idea is that we need to add proper types to metabase-lib, cleanup code a bit, and remove ts-nocheck. I'll outline in the RFC

File: frontend/src/metabase-lib/lib/metadata/metadata.d.ts (hunk 3 above)
Comment (ranquild) [reply frontend/src/metabase-lib/lib/metadata/metadata.d.ts:2]: And it's not needed

**Review by alxnddr — APPROVED** [review]

**Pull request merged**